internal/controllers/fake/client.go

```diff
@@ -0,0 +1,52 @@
+package fake
+
+import (
+	"context"
+	"encoding/json"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Wrap the fake client in other assertions we want to make.
+type fakeTiltClient struct {
+	ctrlclient.Client
+}
+
+// NOTE(nick): We've had a lot of bugs due to the way the apiserver
+// modifies and truncates objects on update.
+//
+// We simulate this behavior to catch this class of bug.
+func (f fakeTiltClient) Update(ctx context.Context, obj client.Object, opts ...client.UpdateOption) error {
```

Comment:
Oh nice! I was looking through the fake client to see if it had any useful hooks (spoiler: it doesn't) but this is a good solution

Reply:
Yaaaaaah. I think the k8s team has put most of their effort into making it easier to use a real api server in tests, rather than making the fake client a good simulation of what the api server does.

But to make that work out, you also need to buy into a bunch of testing philosophies that our tests don't currently buy into... (ginkgo's per-package setup, ways to clear the server between tests, etc)

We might have to end up investing in our own fake client...